File: packages/editing/src/event-transactions.ts
import {
  ABI_VERSION,
  EVENT_TRANSACTIONS_MAGIC,
  EVENT_TRANSACTION_LAYOUTS,
  EventTransactionOpcode,
  INSTRUCTION_HEADER_BYTES,
  MAX_EVENT_TRANSACTIONS_BYTES,
  MAX_EVENT_TRANSACTION_INSTRUCTIONS,
  MAX_RESOURCE_BYTES,
  PROTOCOL_ALIGNMENT,
  STREAM_HEADER_BYTES,
} from "./generated";
import type { InputEventKind } from "./input-stream";

export interface EventTransaction {
  readonly eventId: number;
  readonly kind: InputEventKind;
  readonly target: number;
  readonly x: number;
  readonly y: number;
  readonly deltaX: number;
  readonly deltaY: number;
  readonly buttons: number;
  readonly modifiers: number;
  readonly pointerId: number;
  readonly elapsedMicros: number;
  readonly path: readonly number[];
}

export class EventTransactionDecodingError extends Error {
  public constructor(message: string) {
    super(message);
    this.name = "EventTransactionDecodingError";
  }
}

/** Validates a complete Core hit-result batch before exposing any callback path. */
export function decodeEventTransactionBatch(input: Uint8Array): readonly EventTransaction[] {
  if (input.byteLength > MAX_EVENT_TRANSACTIONS_BYTES)
    fail("event transaction stream is too large");
  if (input.byteLength % PROTOCOL_ALIGNMENT !== 0) fail("event transaction stream is not aligned");
  const reader = new Reader(input);
  if (reader.u32() !== EVENT_TRANSACTIONS_MAGIC) fail("wrong event transaction magic");
  if (reader.u16() !== ABI_VERSION) fail("unsupported event transaction ABI version");
  if (reader.u16() !== STREAM_HEADER_BYTES) fail("invalid event transaction header length");
  if (reader.u32() !== input.byteLength) fail("event transaction length does not match input");
  const declared = reader.u32();
  if (declared > MAX_EVENT_TRANSACTION_INSTRUCTIONS) {
    fail("event transaction instruction count exceeds limit");
  }
  if (declared > Math.floor(reader.remaining / EVENT_TRANSACTION_LAYOUTS[1].minimumBytes)) {
    fail("event transaction instruction count cannot fit in input");
  }
  const events: EventTransaction[] = [];
  while (reader.remaining > 0) {
    const offset = reader.offset;
    const opcode = reader.instruction();
    if (opcode !== EventTransactionOpcode.Event) fail("unknown event transaction opcode");
    events.push(decodeEvent(reader));
    const consumed = reader.offset - offset;
    if (consumed < EVENT_TRANSACTION_LAYOUTS[opcode].minimumBytes || consumed % 4 !== 0) {
      fail("event transaction instruction has invalid length");
    }
  }
  if (events.length !== declared) fail("event transaction count does not match input");
  return events;
}

function decodeEvent(reader: Reader): EventTransaction {
  const eventId = reader.u32();
  const kind = eventKind(reader.u16());
  reader.zeroes(2);
  const target = reader.u32();
  const x = reader.f32();
  const y = reader.f32();
  const deltaX = reader.f32();
  const deltaY = reader.f32();
  const buttons = reader.u32();
  const modifiers = reader.u32();
  const pointerId = reader.u32();
  const elapsedMicros = reader.u32();
  const pathCount = reader.u32();
  if (
    pathCount > Math.floor(MAX_RESOURCE_BYTES / 4) ||
    pathCount > Math.floor(reader.remaining / 4)
  ) {
    fail("event transaction path exceeds its byte envelope");
  }
  const path = Array.from({ length: pathCount }, () => reader.u32());
  if (path.length === 0 || target === 0xffff_ffff || path.at(-1) !== target) {
    fail("event transaction path does not end at its target");
  }
  if (new Set(path).size !== path.length || path.includes(0xffff_ffff)) {
    fail("event transaction path contains a null or repeated node");
  }
  if (buttons > 0xffff || modifiers > 0x0f) fail("event transaction flag bits are reserved");
  if (elapsedMicros < 1 || elapsedMicros > 1_000_000) {
    fail("event transaction elapsed time is invalid");
  }
  return {
    eventId,
    kind,
    target,
    x,
    y,
    deltaX,
    deltaY,
    buttons,
    modifiers,
    pointerId,
    elapsedMicros,
    path,
  };
}

function eventKind(value: number): InputEventKind {
  switch (value) {
    case 1:
      return "pointerdown";
    case 2:
      return "pointerup";
    case 3:
      return "pointermove";
    case 4:
      return "pointercancel";
    case 5:
      return "click";
    case 6:
      return "wheel";
    default:
      return fail("unknown event transaction kind");
  }
}

class Reader {
  readonly #bytes: Uint8Array;
  readonly #view: DataView;
  #offset = 0;

  public constructor(bytes: Uint8Array) {
    this.#bytes = bytes;
    this.#view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  }

  public get offset(): number {
    return this.#offset;
  }

  public get remaining(): number {
    return this.#bytes.byteLength - this.#offset;
  }

  public instruction(): EventTransactionOpcode {
    if (this.#offset % PROTOCOL_ALIGNMENT !== 0) fail("event transaction instruction is unaligned");
    this.require(INSTRUCTION_HEADER_BYTES);
    const opcode = this.u8();
    if (this.u8() !== 0) fail("event transaction instruction flags are unsupported");
    if (this.u16() !== 0) fail("event transaction reserved bytes must be zero");
    return opcode;
  }

  public u8(): number {
    this.require(1);
    return this.#view.getUint8(this.#offset++);
  }

  public u16(): number {
    this.require(2);
    const value = this.#view.getUint16(this.#offset, true);
    this.#offset += 2;
    return value;
  }

  public u32(): number {
    this.require(4);
    const value = this.#view.getUint32(this.#offset, true);
    this.#offset += 4;
    return value;
  }

  public f32(): number {
    this.require(4);
    const value = this.#view.getFloat32(this.#offset, true);
    this.#offset += 4;
    if (!Number.isFinite(value)) fail("event transaction float is not finite");
    return value;
  }

  public zeroes(length: number): void {
    this.require(length);
    for (let index = 0; index < length; index += 1) {
      if (this.#bytes[this.#offset++] !== 0) fail("event transaction reserved bytes must be zero");
    }
  }

  private require(length: number): void {
    if (!Number.isSafeInteger(length) || length < 0 || length > this.remaining) {
      fail("truncated event transaction stream");
    }
  }
}

function fail(message: string): never {
  throw new EventTransactionDecodingError(message);
}

File: core/doper-abi/src/event_transactions.rs
use std::collections::HashSet;

use crate::codec::{
    Reader, Writer, read_header, read_instruction_header, validate_encode_instruction_count,
};
use crate::{
    AbiError, EVENT_TRANSACTIONS_MAGIC, EventTransactionOpcode, InputEventKind,
    MAX_EVENT_TRANSACTION_INSTRUCTIONS, MAX_EVENT_TRANSACTIONS_BYTES, MAX_RESOURCE_BYTES,
    NULL_NODE_ID, StreamKind,
};

/// One Core-hit-tested event and its stable propagation path.
#[derive(Clone, Debug, PartialEq)]
pub struct EventTransactionRecord {
    /// Host-monotonic event identifier.
    pub event_id: u32,
    /// Browser-independent event category.
    pub kind: InputEventKind,
    /// Deepest hit node.
    pub target: u32,
    /// Canvas-local logical coordinates.
    pub position: [f32; 2],
    /// Wheel delta or zeroes.
    pub delta: [f32; 2],
    /// Browser pointer button bitset.
    pub buttons: u32,
    /// Shift/Control/Alt/Meta bits.
    pub modifiers: u32,
    /// Browser pointer identity, or zero for non-pointer events.
    pub pointer_id: u32,
    /// Time since the previous related sample in microseconds.
    pub elapsed_micros: u32,
    /// Root-to-target generation-bearing path.
    pub path: Vec<u32>,
}

/// Transactional reverse-event batch.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct EventTransactionBatch {
    /// Events in accepted input order.
    pub records: Vec<EventTransactionRecord>,
}

impl EventTransactionBatch {
    /// Fully decodes and validates untrusted reverse event bytes.
    pub fn decode(bytes: &[u8]) -> Result<Self, AbiError> {
        let mut reader = Reader::new(bytes);
        let declared = read_header(
            &mut reader,
            EVENT_TRANSACTIONS_MAGIC,
            MAX_EVENT_TRANSACTIONS_BYTES,
        )?;
        if declared > MAX_EVENT_TRANSACTION_INSTRUCTIONS {
            return Err(AbiError::InstructionCountTooLarge {
                declared,
                maximum: MAX_EVENT_TRANSACTION_INSTRUCTIONS,
            });
        }
        let minimum = EventTransactionOpcode::Event.minimum_bytes();
        let maximum = u32::try_from(reader.remaining() / minimum)
            .map_err(|_| AbiError::ArithmeticOverflow)?;
        if declared > maximum {
            return Err(AbiError::InstructionCountTooLarge { declared, maximum });
        }
        let capacity = usize::try_from(declared).map_err(|_| AbiError::ArithmeticOverflow)?;
        let mut records = Vec::with_capacity(capacity);
        while reader.remaining() > 0 {
            let (offset, raw_opcode, _) = read_instruction_header(&mut reader)?;
            let opcode =
                EventTransactionOpcode::from_u8(raw_opcode).ok_or(AbiError::UnknownOpcode {
                    stream: StreamKind::EventTransactions,
                    opcode: raw_opcode,
                    offset,
                })?;
            let record = decode_record(&mut reader)?;
            validate_size(opcode, offset, reader.offset())?;
            records.push(record);
        }
        let actual = u32::try_from(records.len()).map_err(|_| AbiError::ArithmeticOverflow)?;
        if actual != declared {
            return Err(AbiError::InstructionCountMismatch { declared, actual });
        }
        Ok(Self { records })
    }

    /// Encodes a canonical aligned reverse event batch.
    pub fn encode(&self) -> Result<Vec<u8>, AbiError> {
        validate_encode_instruction_count(
            self.records.len(),
            0,
            MAX_EVENT_TRANSACTION_INSTRUCTIONS,
        )?;
        let mut writer = Writer::new(EVENT_TRANSACTIONS_MAGIC);
        for record in &self.records {
            validate_record(record)?;
            let offset = writer.offset();
            writer.instruction(EventTransactionOpcode::Event as u8, 0);
            writer.u32(record.event_id);
            writer.u16(record.kind as u16);
            writer.u16(0);
            writer.u32(record.target);
            for value in record.position.into_iter().chain(record.delta) {
                writer.f32(value)?;
            }
            writer.u32(record.buttons);
            writer.u32(record.modifiers);
            writer.u32(record.pointer_id);
            writer.u32(record.elapsed_micros);
            writer.u32(u32::try_from(record.path.len()).map_err(|_| AbiError::ArithmeticOverflow)?);
            for node in &record.path {
                writer.u32(*node);
            }
            validate_size(EventTransactionOpcode::Event, offset, writer.offset())?;
        }
        writer.finish(MAX_EVENT_TRANSACTIONS_BYTES)
    }
}

fn decode_record(reader: &mut Reader<'_>) -> Result<EventTransactionRecord, AbiError> {
    let event_id = reader.read_u32()?;
    let kind = InputEventKind::decode(reader.read_u16()?)?;
    reader.read_zeroes(2)?;
    let target = reader.read_u32()?;
    let position = [reader.read_f32()?, reader.read_f32()?];
    let delta = [reader.read_f32()?, reader.read_f32()?];
    let buttons = reader.read_u32()?;
    let modifiers = reader.read_u32()?;
    let pointer_id = reader.read_u32()?;
    let elapsed_micros = reader.read_u32()?;
    let path_count =
        usize::try_from(reader.read_u32()?).map_err(|_| AbiError::ArithmeticOverflow)?;
    let path_bytes = path_count
        .checked_mul(4)
        .ok_or(AbiError::ArithmeticOverflow)?;
    if path_bytes > MAX_RESOURCE_BYTES || path_bytes > reader.remaining() {
        return Err(AbiError::ResourceTooLarge {
            actual: path_bytes,
            maximum: MAX_RESOURCE_BYTES.min(reader.remaining()),
        });
    }
    let mut path = Vec::with_capacity(path_count);
    for _ in 0..path_count {
        path.push(reader.read_u32()?);
    }
    let record = EventTransactionRecord {
        event_id,
        kind,
        target,
        position,
        delta,
        buttons,
        modifiers,
        pointer_id,
        elapsed_micros,
        path,
    };
    validate_record(&record)?;
    Ok(record)
}

fn validate_record(record: &EventTransactionRecord) -> Result<(), AbiError> {
    if record.path.is_empty()
        || record.target == NULL_NODE_ID
        || record.path.last() != Some(&record.target)
    {
        return Err(AbiError::InvalidValue(
            "event path must end at a non-null target",
        ));
    }
    if record.path.len().saturating_mul(4) > MAX_RESOURCE_BYTES {
        return Err(AbiError::ResourceTooLarge {
            actual: record.path.len().saturating_mul(4),
            maximum: MAX_RESOURCE_BYTES,
        });
    }
    let mut unique = HashSet::with_capacity(record.path.len());
    if record
        .path
        .iter()
        .any(|node| *node == NULL_NODE_ID || !unique.insert(*node))
    {
        return Err(AbiError::InvalidValue(
            "event path contains a null or repeated node",
        ));
    }
    if record.buttons & !0xffff != 0 || record.modifiers & !0x0f != 0 {
        return Err(AbiError::InvalidValue(
            "event button or modifier bits are reserved",
        ));
    }
    if record.elapsed_micros == 0 || record.elapsed_micros > 1_000_000 {
        return Err(AbiError::InvalidValue("event elapsed time is invalid"));
    }
    Ok(())
}

fn validate_size(
    opcode: EventTransactionOpcode,
    offset: usize,
    end: usize,
) -> Result<(), AbiError> {
    let actual = end
        .checked_sub(offset)
        .ok_or(AbiError::ArithmeticOverflow)?;
    if actual < opcode.minimum_bytes() || actual % crate::PROTOCOL_ALIGNMENT != 0 {
        return Err(AbiError::InstructionLengthMismatch {
            opcode: opcode as u8,
            offset,
            expected: opcode.minimum_bytes(),
            actual,
        });
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn event_path_round_trips_and_rejects_cycles() {
        let record = EventTransactionRecord {
            event_id: 9,
            kind: InputEventKind::PointerDown,
            target: 3,
            position: [12.0, 20.0],
            delta: [0.0, 0.0],
            buttons: 1,
            modifiers: 4,
            pointer_id: 1,
            elapsed_micros: 16_667,
            path: vec![1, 2, 3],
        };
        let batch = EventTransactionBatch {
            records: vec![record],
        };
        let bytes = batch.encode().expect("encode");
        assert_eq!(EventTransactionBatch::decode(&bytes), Ok(batch));
        let cyclic = EventTransactionBatch {
            records: vec![EventTransactionRecord {
                path: vec![1, 1],
                target: 1,
                ..EventTransactionRecord {
                    event_id: 1,
                    kind: InputEventKind::Click,
                    target: 1,
                    position: [0.0, 0.0],
                    delta: [0.0, 0.0],
                    buttons: 0,
                    modifiers: 0,
                    pointer_id: 0,
                    elapsed_micros: 1,
                    path: vec![],
                }
            }],
        };
        assert!(cyclic.encode().is_err());
    }

    #[test]
    fn hostile_event_records_fail_closed_on_every_validated_field() {
        let valid = EventTransactionRecord {
            event_id: 9,
            kind: InputEventKind::PointerMove,
            target: 3,
            position: [12.0, 20.0],
            delta: [0.0, 0.0],
            buttons: 1,
            modifiers: 4,
            pointer_id: 1,
            elapsed_micros: 16_667,
            path: vec![1, 2, 3],
        };
        let reject = |mutate: fn(&mut EventTransactionRecord)| {
            let mut record = valid.clone();
            mutate(&mut record);
            assert!(
                EventTransactionBatch {
                    records: vec![record],
                }
                .encode()
                .is_err()
            );
        };
        reject(|record| record.path.clear());
        reject(|record| record.target = crate::NULL_NODE_ID);
        reject(|record| record.path = vec![1, 2]);
        reject(|record| record.path = vec![crate::NULL_NODE_ID, 3]);
        reject(|record| record.buttons = 0x1_0000);
        reject(|record| record.modifiers = 0x10);
        reject(|record| record.elapsed_micros = 0);
        reject(|record| record.elapsed_micros = 2_000_000);

        let bytes = EventTransactionBatch {
            records: vec![valid],
        }
        .encode()
        .expect("encode");
        // Truncations and every byte-flip either decode or fail without panic.
        for cut in 0..bytes.len() {
            let _ = EventTransactionBatch::decode(&bytes[..cut]);
        }
        for index in 0..bytes.len() {
            let mut hostile = bytes.clone();
            hostile[index] ^= 0xff;
            let _ = EventTransactionBatch::decode(&hostile);
        }
        let mut wrong_count = bytes.clone();
        wrong_count[12..16].copy_from_slice(&999_u32.to_le_bytes());
        assert!(EventTransactionBatch::decode(&wrong_count).is_err());
    }
}
